notebooks/notebook_template_review.py

```diff
@@ -426,495 +515,446 @@ def validate(self, path: str, cells: list) -> None:
 
+class SentenceCaseTWRule(TextRule):
+    def validate(self,
+                 notebook,
+                 text: List[str]) -> bool:
+        """
+        Check that headings are in sentence case
 
-def get_cell(path: str, 
-             cells: list, 
-             cell_index: int) -> (list, int):
-    """
-        Get the next notebook cell.
-        
         path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
-        cell_index: The index of the last cell that was parsed (reviewed).
-        
-        Returns:
-        
-        cell: content of the next cell
-        cell_index + 1: index of subsequent cell
-    """
-    while empty_cell(path, cells, cell_index):
-        cell_index += 1
-        
-    cell = cells[cell_index]
-    if cell['cell_type'] == 'markdown':
-        check_text_cell(path, cell)
-    return cell, cell_index + 1
+        text: the heading to check
+        """
+        ret = True
 
+        ACRONYMS = ['E2E', 'Vertex', 'AutoML', 'ML', 'AI', 'GCP', 'API', 'R', 'CMEK', 
+                    'TF', 'TFX', 'TFDV', 'SDK', 'VM', 'CPR', 'NVIDIA', 'ID', 'DASK', 
+                    'ARIMA_PLUS', 'KFP', 'I/O', 'GPU', 'Google', 'TensorFlow', 'PyTorch'
+                    ]
 
-def empty_cell(path: str, 
-               cells: list, 
-               cell_index: int) -> bool:
-    """
-        Check for empty cells
-        
-        path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
-        cell_index: The index of the last cell that was parsed (reviewed).
-        
-        Returns:
-        
-        bool: whether cell is empty or not
-    """
-    if len(cells[cell_index]['source']) == 0:
-        report_error(path, ErrorCode.ERROR_EMPTY_CELL, f'empty cell: cell #{cell_index}')
-        return True
-    else:
-        return False
+        # Check the first line
+        words = text[0].replace('#', '').split(' ')
+        if not words[0][0].isupper():
+            ret = notebook.report_error(ErrorCode.ERROR_HEADING_CAP, f"heading must start with capitalized word: {words[0]}")
 
+        for word in words[1:]:
+            word = word.replace(':', '').replace('(', '').replace(')', '')
+            if word in ACRONYMS:
+                continue
+            if word.isupper():
+                ret = notebook.report_error(ErrorCode.ERROR_HEADING_CASE, f"heading is not sentence case: {word}")
+                
+        return ret
 
-def report_error(notebook: str, 
-                 code: ErrorCode,
-                 errmsg: str) -> None:
-    """
-    Report an error.
-        If args.errors_codes set, then only report these errors. Otherwise, all errors.
-    
-    notebook: The notebook path.
-    code: The error code number.
-    errmsg: The error message
-    """
-    
-    global num_errors
+
+
+class TextTWRule(TextRule):
+    def validate(self, notebook: Notebook, text: List[str]) -> bool:
```

Comment:
Signature not conforming to abstractclass.

Reply:
That means that this method is never called and your abstract classes default method is called instead.

Reply:
This is derived from the TextRule abstract class, not the general NotebookRule class. It assumes that instead of passing a cell, it is passed some passage (list) of text strings.